src/tir/schedule/primitive/blockize_tensorize.cc

```diff
@@ -0,0 +1,638 @@
+    } else if (!outer && !inner) {
+      const auto& outer_iter = unit_iter_mark;
+      const auto& inner_iter = unit_iter_mark;
+      res.push_back({outer_iter, inner_iter});
+    } else {
+      return {};
+    }
+  }
+  res.push_back({arith::IterMark(arith::IterSumExpr({}, 0), Bool(true)),
+                 arith::IterMark(arith::IterSumExpr({}, 0), Bool(true))});
+  return res;
+}
+
+/*!
+ * \brief Generate the blockized init block.
+ * \param block The original block with init.
+ * \param inner_block_realize The block realize of the inner block after blockize.
+ * \param inner_loops The inner loops after blockize.
+ * \return The subtree of the init block and its outer loops.
+ */
+Stmt GenerateBlockizedInit(const Block& block, const BlockRealize& inner_block_realize,
+                           const std::vector<const ForNode*>& inner_loops) {
+  Array<IterVar> init_block_iters;
+  Array<PrimExpr> init_bindings;
+  const Block& inner_block = inner_block_realize->block;
+
+  // Step 1: Collect data-parallel block iters
+  for (size_t i = 0; i < inner_block->iter_vars.size(); i++) {
+    const IterVar& iter_var = inner_block->iter_vars[i];
+    const PrimExpr& binding = inner_block_realize->iter_values[i];
+    if (iter_var->iter_type == IterVarType::kDataPar &&
+        UsesVar(block->init.value(),
+                [&iter_var](const VarNode* var) { return var == iter_var->var.get(); })) {
+      init_block_iters.push_back(iter_var);
+      init_bindings.push_back(binding);
+    }
+  }
+
+  // Step 2: Collect loops related to iters of the init block
+  std::vector<const ForNode*> init_loops;
+  for (const ForNode* inner_loop : inner_loops) {
+    for (const PrimExpr& init_binding : init_bindings) {
+      if (UsesVar(init_binding,
+                  [inner_loop](const VarNode* var) { return var == inner_loop->loop_var.get(); })) {
+        init_loops.push_back(inner_loop);
+      }
+    }
+  }
+
+  // Step 3: Create new block iters for the init block
+  Map<Var, PrimExpr> subst_map;
+  for (size_t i = 0; i < init_block_iters.size(); i++) {
+    IterVar new_iter_var = init_block_iters[i];
+    auto* new_init_var_node = new_iter_var.CopyOnWrite();
+    Var old_var = new_iter_var->var;
+    new_init_var_node->var = old_var.copy_with_suffix("_init");
+    subst_map.Set(old_var, new_iter_var->var);
+    init_block_iters.Set(i, std::move(new_iter_var));
+  }
+
+  // Step 4: Generate loop nests and the init block
+  Block init_block{/*iter_vars=*/init_block_iters,            //
+                   /*reads=*/{},                              //
+                   /*writes=*/block->writes,                  //
+                   /*name_hint=*/block->name_hint + "_init",  //
+                   /*body=*/block->init.value(),              //
+                   /*init=*/NullOpt};
+  Stmt new_init = BlockRealize(
+      /*iter_values=*/init_bindings,
+      /*predicate=*/inner_block_realize->predicate,
+      /*block=*/std::move(init_block));
```

Comment:
nit: can save an intermediate variable

```suggestion
  Stmt new_init = BlockRealize(
      /*iter_values=*/init_bindings,
      /*predicate=*/inner_block_realize->predicate,
      /*block=*/
      Block{/*iter_vars=*/init_block_iters,
            /*reads=*/{},
            /*writes=*/block->writes,
            /*name_hint=*/block->name_hint + "_init",
            /*body=*/block->init.value(),
            /*init=*/NullOpt});
```